pyccel/ast/numpyext.py

```diff
@@ -151,6 +151,80 @@ def fprint(self, printer, lhs=None):
             return '{0} = sum({1})'.format(lhs_code, rhs_code)
         return 'sum({0})'.format(rhs_code)
 
+#=======================================================================================
+
+class Product(Function):
+    """Represents a call to  numpy.prod for code generation.
+
+    arg : list , tuple , Tuple, List, Variable
+    """
+
+    def __new__(cls, arg):
+        if not isinstance(arg, (list, tuple, Tuple, List, Variable, Mul, Add, Pow, sp_Rational)):
+            raise TypeError('Uknown type of  %s.' % type(arg))
+        return Basic.__new__(cls, arg)
+
+    @property
+    def arg(self):
+        return self._args[0]
+
+    @property
+    def dtype(self):
+        return self._args[0].dtype
+
+    @property
+    def rank(self):
+        return 0
+
+    def fprint(self, printer, lhs=None):
+        """Fortran print."""
+
+        rhs_code = printer(self.arg)
+        if lhs:
+            lhs_code = printer(lhs)
+            return '{0} = product({1})'.format(lhs_code, rhs_code)
+        return 'product({0})'.format(rhs_code)
+
+#=======================================================================================
+
+# TODO: Remove hack to get order right with Fortran indexing
```

Comment:
Once this change is made we can merge this branch into master.

Reply:
Interesting. `fprint` never gets called, even when adding an additional entry in `fcode.f90`, and it's listed as "remove fprint from ast classes (see numpyext)" in TODO. In the end `Matmul(A,x)` is printed with large `M` and this order, no matter what I do. @ratnania, any ideas?

Reply:
Resolved. One must not define `Matmul` of base class `Function` but rather only `Application`. Now the swapping of arguments works correctly. There is probably a lot of dead code in `fprint` inside `numpyext.py`, as these types are all defined via `Function` and just get caught before this is ever called.

Reply:
Yes, this is very interesting indeed. I did know about that TODO item. In the case of `Array` and `Zeros` I made some changes to the `fprint` method as suggested by @saidctb and they now work as we want in branch `devel-numpyext`.

Reply:
I've now added support for both orderings F and C and this should work fine with matmul now according to test cases. Should be ready to merge in my opinion. Anyway, `Function` vs `Application` and `fprint` should be fixed independently.

Reply:
Function/Application was added in #237 branch by @saidctb , so everything should be fine with this pull request. Since this branch #235 branches from `master` anyway, we can merge it. @yguclu , can you do that?